[pipelining] rewrite interleaved 1f1b

H-Huang opened the pull request:
## Context

Interleaved 1F1B has multiple points in the schedule where communication is both criss-crossed across ranks leading to hangs due to 1. looped nature of schedules, 2. batched nature of forward + backward in 1f1b phase.

<img width="1370" alt="image" src="https://github.com/pytorch/pytorch/assets/14858254/a07c2b1d-8a99-420b-9ba3-32a0115d228b">

In the current implementation, it is difficult to fix these hangs since it requires `dist.recv` from a prior point in time, but each rank operates on its own step schedule and does not have knowledge of other ranks operations to perform the `recv` prior to their own `send`.

## New implementation

The new implementation is split into 2 parts:

1. Creating the pipeline order.

Each rank will create the timestep normalized ordering of all schedule actions across all ranks. This is created once during the initialization of the schedule class. The timestep between each rank is normalized as each rank can only have 1 computation action (forward or backward) during that timestep.

<img width="1065" alt="image" src="https://github.com/pytorch/pytorch/assets/14858254/196f2347-7ff4-49cf-903b-d8db97d1156f">

3. Executing the pipeline order.

Once the pipeline order is determined, execution is simple because as each rank will perform its send to its peer (based on whether they did forward and backward). Now that each rank has a global understanding of the schedule, they can check their previous and next neighbor ranks to see if they need to recv any activations/gradients from them. Therefore, during execution, each rank is aligned and executing the same time step.

## Benefits

- Implementation is faster since 1f1b computation can now be split up in two time steps, 1 for forward and 1 for backward.
- Debugging is easier since we can now determine which timestep each rank is hung on
- Testing is easier since we can just validate the pipeline order, without running the schedule. This allows us to test on large amount of ranks without actually needing the GPUs.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #127796
* #127559
* __->__ #127332
* #127084



cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

H-Huang committed:
[pipelining] rewrite interleaved 1f1b

[ghstack-poisoned]

H-Huang committed:
Update on "[pipelining] rewrite interleaved 1f1b"

## Context

Interleaved 1F1B has multiple points in the schedule where communication is both criss-crossed across ranks leading to hangs due to 1. looped nature of schedules, 2. batched nature of forward + backward in 1f1b phase.

<img width="1370" alt="image" src="https://github.com/pytorch/pytorch/assets/14858254/a07c2b1d-8a99-420b-9ba3-32a0115d228b">

In the current implementation, it is difficult to fix these hangs since it requires `dist.recv` from a prior point in time, but each rank operates on its own step schedule and does not have knowledge of other ranks operations to perform the `recv` prior to their own `send`.

## New implementation

The new implementation is split into 2 parts:

1. Creating the pipeline order.

Each rank will create the timestep normalized ordering of all schedule actions across all ranks. This is created once during the initialization of the schedule class. The timestep between each rank is normalized as each rank can only have 1 computation action (forward or backward) during that timestep.

<img width="1065" alt="image" src="https://github.com/pytorch/pytorch/assets/14858254/196f2347-7ff4-49cf-903b-d8db97d1156f">

3. Executing the pipeline order.

Once the pipeline order is determined, execution is simple because as each rank will perform its send to its peer (based on whether they did forward and backward). Now that each rank has a global understanding of the schedule, they can check their previous and next neighbor ranks to see if they need to recv any activations/gradients from them. Therefore, during execution, each rank is aligned and executing the same time step.

## Benefits

- Implementation is faster since 1f1b computation can now be split up in two time steps, 1 for forward and 1 for backward.
- Debugging is easier since we can now determine which timestep each rank is hung on
- Testing is easier since we can just validate the pipeline order, without running the schedule. This allows us to test on large amount of ranks without actually needing the GPUs.




cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k 

[ghstack-poisoned]

kwen2501 commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -652,14 +658,6 @@ def _step_microbatches(
 
 
 class ScheduleInterleaved1F1B(PipelineScheduleMulti):
-    """
-    The Interleaved 1F1B schedule.
-    Will perform one forward and one backward on the microbatches in steady
-    state and supports multiple stages per rank. When microbatches are ready for
-    multiple local stages, Interleaved 1F1B prioritizes the earlier microbatch
-    (also called "depth first").
-    """
-
```
Replacement for this documentation?

H-Huang replied:
Accidentally removed it, I added it back in with additional comments

kwen2501 commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -685,33 +682,29 @@ def __init__(
+
+        # 1. Create the pipeline_order (all ranks do this calculation)
+        # This will be used to keep track of the current state of the entire pipeline
+        # pipeline_order[rank] = [(computation_type, microbatch_index, stage_index), ...]
+        self.pipeline_order: List[
+            List[Optional[Tuple[_ComputationType, int, int]]]
+        ] = []
```
> pipeline_order[rank] = [(computation_type, microbatch_index, stage_index), ...]

For the left-hand side, use dict with `rank` as key, rather than using list.
For the right-hand side, document what the `List` structure is for.

H-Huang replied:
rank is local_rank so it will always be 0 to group_size. But I will update this and also update the tuple to a namedtuple.

kwen2501 commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -685,33 +682,29 @@ def __init__(
+        self, rank
+    ) -> List[Optional[Tuple[_ComputationType, int, int]]]:
+        def get_rank_warmup_steps(rank):
+            # Increment warmup_steps by 2 for each hop away
+            warmup_steps = (self.n_local_stages - 1) * self.pp_group_size
+            warmup_steps += 2 * ((self.pp_group_size - 1) - rank)
```
nit: comment on the whole calculation. The current comment seems to apply to the second line of code only.

wconstab replied:
yea, better comment would help.  but dont just tell us literally what the math is doing, try to tell us the reasoning.  and perhaps use more temporary variables with instructive names instead of adding comments if that makes it clearer.

The first line `warmup_steps = ` - what is the logical meaning of this line- the number of warmup steps for the last pp stage?  if so, maybe name it `warmups_last_stage`

also, wouldn't the last stage not have any warmups? or im confused

H-Huang replied:
Added comments and renamed variables

wconstab replied:
i see. its actually not clear to me why it makes sense to calcuate num_warmups per rank instead of per stage.  wouldn't each stage have a different number of warmups on any given rank?

the variable names used to be 'stage' based, which actually does seem logical to me

H-Huang replied:
Previously we were operating at rank level, each rank does determine which stage it will operate on based on the timestep. The timestep should be at rank level since communication is also done at the rank level. I renamed the "steps" variable to be "ops" since operations is more accurate.

kwen2501 commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -685,33 +682,29 @@ def __init__(
+            warmup_steps += 2 * ((self.pp_group_size - 1) - rank)
+            return min(warmup_steps, self._n_microbatches * self.n_local_stages)
+
+        warmup_steps = get_rank_warmup_steps(rank)
         fwd_bwd_steps = (self.n_local_stages * self._n_microbatches) - warmup_steps
         cooldown_steps = (self.n_local_stages * self._n_microbatches) - fwd_bwd_steps
```
The PR says "steps" are normalized now, meaning each step can be only 1 forward or 1 backward. Are the `*_steps` here normalized now? Asking because of the `assert` below:
```
        assert (
            warmup_steps + fwd_bwd_steps * 2 + cooldown_steps
            == self.n_local_stages * self._n_microbatches * 2
        )
```
There is a `fwd_bwd_steps * 2`

H-Huang replied:
There "steps" are not normalized, so I updated the variable names to "warmup_ops", "1f1b_ops", etc. instead of steps to make it more clear

kwen2501 commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -722,133 +715,172 @@ def _step_microbatches(
             )
+            return (local_index * self.pp_group_size) + rank
+
+        fwd_stage_mb_index: Dict[int, int] = defaultdict(int)
+        bwd_stage_mb_index: Dict[int, int] = defaultdict(int)
```
nit: document the meaning of the `key`

kwen2501 replied:
nit: `defaultdict` is perhaps not a good fit here -- we don't want a `0` value (a valid mb_index) being assigned by unknown reason.

kwen2501 commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -722,133 +715,172 @@ def _step_microbatches(
+                    stage = stage_index_to_stage[stage_index]
+                    loss = self._maybe_get_loss(stage, mb_index)
+                    stage.backward_one_chunk(loss=loss)
+                    if stage_index == 4 and mb_index == 4:
+                        pass
```
Unclean code?

H-Huang replied:
thanks, accidentally left that in when debugging

kwen2501 commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -722,133 +715,172 @@ def _step_microbatches(
+                    raise ValueError(f"Unknown computation type {computation_type}")
+
+            # look at neighboring ranks to see if I need to include any recv communication
+            if time_step < len(prev_rank_ops):
```
nit: what does the `if` mean here?

H-Huang replied:
Added comments, but this is needed since the some ranks may be done with their computation earlier than other ranks, as a result their total number of timesteps will be shorter.

wconstab replied:
actually now im confused, i thought we were adding None actions to essentially pad each rank so their timesteps were the same.  Is that only true for the warmup side, but not the cooldown side?

H-Huang replied:
the None actions were padded after warmup such that the 1f1bs are aligned, then we do pad during cooldown (https://github.com/pytorch/pytorch/pull/127332/files#diff-bcea404838585952f926a03a7da41b33f8bdc541a955ef25345a69ab2e2a286eR809-R811) to make sure that backwards are also aligned

kwen2501 commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -722,133 +715,172 @@ def _step_microbatches(
+                # only handle sends for the forward from a previous rank
+                if computation_type == _ComputationType.FORWARD:
+                    if stage_index != self._num_stages - 1:
+                        stage = stage_index_to_stage[stage_index + 1]
```
Note: this line is assuming that `stage_index + 1` is the receiver of `stage_index`. 
This is correct for the most cases, but also overrides the comm relationship returned by `stage.get_fwd_recv_ops()` (which will return the source rank based on pipe construction).

We should at least document the assumption here.

kwen2501 commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -722,133 +715,172 @@ def _step_microbatches(
+                    if stage_index != self._num_stages - 1:
+                        stage = stage_index_to_stage[stage_index + 1]
+                        ops.extend(stage.get_fwd_recv_ops())
+                elif computation_type == _ComputationType.BACKWARD:
+                    pass
```
Please document the reason for the "pass" here. In particular, why it is safe to pass.

wconstab replied:
prev-rank doing backward has no influence on us.  prev-rank doing forward may influence us to do a recv.  (next-rank doing backward similarly).

kwen2501 commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -722,133 +715,172 @@ def _step_microbatches(
+            if next_rank_action is not None:
+                computation_type, mb_index, stage_index = next_rank_action
+                # only handle receives for the backwards from a next rank
+                if computation_type == _ComputationType.FORWARD:
+                    pass
```
Same.

kwen2501 commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -722,133 +715,172 @@ def _step_microbatches(
+                    pass
+                elif computation_type == _ComputationType.BACKWARD:
+                    if stage_index != 0:
+                        stage = stage_index_to_stage[stage_index - 1]
+                        ops.extend(stage.get_bwd_recv_ops())
```
Same.

wconstab commented:
> Implementation is faster since 1f1b computation can now be split up in two time steps, 1 for forward and 1 for backward.

This isn't really clear to me- the splitting of time-steps from (F+B) in a step to (F) or (B) seems purely logical to me.  It should be just as fast either way, shouldn't it?

wconstab commented:
> Debugging is easier since we can now determine which timestep each rank is hung on

Not sure if i'm missing something, but we knew which timestep a rank was on before too didn't we?  (if we used a debugger or log messages).  Isn't the only difference now that perhaps its a bit easier to reason about the schedule if you can see it written out step by step?

>Testing is easier since we can just validate the pipeline order, without running the schedule. This allows us to test on large amount of ranks without actually needing the GPUs.

I think testing on real gpus will always be important as final validation, even if this approach gives us an _extra_ place to do up front unit tests on the schedule logic.  That is a definite benefit

wconstab commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -722,133 +715,172 @@ def _step_microbatches(
+        for _ in range(rank):
+            rank_ops.append(None)
+
+        # These is used to calculate the number of slots to fill with no-ops, to account for the delay in warmup
```
These are

wconstab commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -722,133 +715,172 @@ def _step_microbatches(
+        # These is used to calculate the number of slots to fill with no-ops, to account for the delay in warmup
+        # when we want to wait for the backward to trickle back up
+        # first backward is going to be pp_group_size - 1 hops away,
+        post_warmup_steps = 2 if rank == 0 else 0
```
confused about this logic and the comment.

logic: why is rank 0 'special'?
comment: where does `pp_group_size - 1` actually factor in?

H-Huang replied:
Sorry the comment was outdated, I added the formula I used in the calculation in the comments.

wconstab commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -722,133 +715,172 @@ def _step_microbatches(
+                rank_ops.append((_ComputationType.FORWARD, mb_index, fwd_stage_index))
+                if step == warmup_steps - 1:
+                    # This is the last step in the warmup phase, so we need to wait for the backward to trickle back up
+                    while post_warmup_steps > 0:
```
`rank_ops.extend([None] * post_warup_steps)`?

H-Huang replied:
thanks!

wconstab commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -722,133 +715,172 @@ def _step_microbatches(
-        # Make sure all sends are finished
-        for work in sends_to_wait:
-            work.wait()
+        TODO: Interleaved 1F1B does not use sorted_batch_isend_irecv(). As a result, this schedule does
```
why is this, actually?

H-Huang replied:
I think this can be reintroduced, especially when we figure out how we want to handle when a pipelinestage has activations to multiple ranks.

H-Huang replied:
create an issue for this

H-Huang replied:
https://github.com/pytorch/pytorch/issues/127964

H-Huang committed:
Update on "[pipelining] rewrite interleaved 1f1b"

## Context

Interleaved 1F1B has multiple points in the schedule where communication is both criss-crossed across ranks leading to hangs due to 1. looped nature of schedules, 2. batched nature of forward + backward in 1f1b phase.

<img width="1370" alt="image" src="https://github.com/pytorch/pytorch/assets/14858254/a07c2b1d-8a99-420b-9ba3-32a0115d228b">

In the current implementation, it is difficult to fix these hangs since it requires `dist.recv` from a prior point in time, but each rank operates on its own step schedule and does not have knowledge of other ranks operations to perform the `recv` prior to their own `send`.

## New implementation

The new implementation is split into 2 parts:

1. Creating the pipeline order.

Each rank will create the timestep normalized ordering of all schedule actions across all ranks. This is created once during the initialization of the schedule class. The timestep between each rank is normalized as each rank can only have 1 computation action (forward or backward) during that timestep.

<img width="1065" alt="image" src="https://github.com/pytorch/pytorch/assets/14858254/196f2347-7ff4-49cf-903b-d8db97d1156f">

3. Executing the pipeline order.

Once the pipeline order is determined, execution is simple because as each rank will perform its send to its peer (based on whether they did forward and backward). Now that each rank has a global understanding of the schedule, they can check their previous and next neighbor ranks to see if they need to recv any activations/gradients from them. Therefore, during execution, each rank is aligned and executing the same time step.

## Benefits

- Implementation is faster since 1f1b computation can now be split up in two time steps, 1 for forward and 1 for backward.
- Debugging is easier since we can now determine which timestep each rank is hung on
- Testing is easier since we can just validate the pipeline order, without running the schedule. This allows us to test on large amount of ranks without actually needing the GPUs.




cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k 

[ghstack-poisoned]

wconstab commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -685,170 +697,221 @@ def __init__(
+        # fwd_bwd_ops should encompass the remaining forwards
+        fwd_bwd_ops = (self.n_local_stages * self._n_microbatches) - warmup_ops
+        # cooldown_ops should encompass the remaining backwards
+        cooldown_ops = (self.n_local_stages * self._n_microbatches) - fwd_bwd_ops
```
maybe rewrite the section as 

```
total_ops = self.n_local_stages * self._n_microbatches   # and move this higher up if its useful in other places
warmup_cooldown_ops = get_rank_warmup_ops
fwd_bwd_ops = total_ops - warmup_cooldown_ops

H-Huang replied:
refactored this!

wconstab commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -685,170 +697,221 @@ def __init__(
+        # twice, once for forward and once for backward
         assert (
-            warmup_steps + fwd_bwd_steps * 2 + cooldown_steps
+            warmup_ops + fwd_bwd_ops * 2 + cooldown_ops
```
you could probably delete all this assertion based on the refactor i suggested above.

H-Huang committed:
Update on "[pipelining] rewrite interleaved 1f1b"

## Context

Interleaved 1F1B has multiple points in the schedule where communication is both criss-crossed across ranks leading to hangs due to 1. looped nature of schedules, 2. batched nature of forward + backward in 1f1b phase.

<img width="1370" alt="image" src="https://github.com/pytorch/pytorch/assets/14858254/a07c2b1d-8a99-420b-9ba3-32a0115d228b">

In the current implementation, it is difficult to fix these hangs since it requires `dist.recv` from a prior point in time, but each rank operates on its own step schedule and does not have knowledge of other ranks operations to perform the `recv` prior to their own `send`.

## New implementation

The new implementation is split into 2 parts:

1. Creating the pipeline order.

Each rank will create the timestep normalized ordering of all schedule actions across all ranks. This is created once during the initialization of the schedule class. The timestep between each rank is normalized as each rank can only have 1 computation action (forward or backward) during that timestep.

<img width="1065" alt="image" src="https://github.com/pytorch/pytorch/assets/14858254/196f2347-7ff4-49cf-903b-d8db97d1156f">

3. Executing the pipeline order.

Once the pipeline order is determined, execution is simple because as each rank will perform its send to its peer (based on whether they did forward and backward). Now that each rank has a global understanding of the schedule, they can check their previous and next neighbor ranks to see if they need to recv any activations/gradients from them. Therefore, during execution, each rank is aligned and executing the same time step.

## Benefits

- Implementation is faster since 1f1b computation can now be split up in two time steps, 1 for forward and 1 for backward.
- Debugging is easier since we can now determine which timestep each rank is hung on
- Testing is easier since we can just validate the pipeline order, without running the schedule. This allows us to test on large amount of ranks without actually needing the GPUs.




cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k 

[ghstack-poisoned]

H-Huang committed:
Update on "[pipelining] rewrite interleaved 1f1b"

## Context

Interleaved 1F1B has multiple points in the schedule where communication is both criss-crossed across ranks leading to hangs due to 1. looped nature of schedules, 2. batched nature of forward + backward in 1f1b phase.

<img width="1370" alt="image" src="https://github.com/pytorch/pytorch/assets/14858254/a07c2b1d-8a99-420b-9ba3-32a0115d228b">

In the current implementation, it is difficult to fix these hangs since it requires `dist.recv` from a prior point in time, but each rank operates on its own step schedule and does not have knowledge of other ranks operations to perform the `recv` prior to their own `send`.

## New implementation

The new implementation is split into 2 parts:

1. Creating the pipeline order.

Each rank will create the timestep normalized ordering of all schedule actions across all ranks. This is created once during the initialization of the schedule class. The timestep between each rank is normalized as each rank can only have 1 computation action (forward or backward) during that timestep.

<img width="1065" alt="image" src="https://github.com/pytorch/pytorch/assets/14858254/196f2347-7ff4-49cf-903b-d8db97d1156f">

3. Executing the pipeline order.

Once the pipeline order is determined, execution is simple because as each rank will perform its send to its peer (based on whether they did forward and backward). Now that each rank has a global understanding of the schedule, they can check their previous and next neighbor ranks to see if they need to recv any activations/gradients from them. Therefore, during execution, each rank is aligned and executing the same time step.

## Benefits

- Implementation is faster since 1f1b computation can now be split up in two time steps, 1 for forward and 1 for backward.
- Debugging is easier since we can now determine which timestep each rank is hung on
- Testing is easier since we can just validate the pipeline order, without running the schedule. This allows us to test on large amount of ranks without actually needing the GPUs.




cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k 

[ghstack-poisoned]

H-Huang committed:
Update on "[pipelining] rewrite interleaved 1f1b"

## Context

Interleaved 1F1B has multiple points in the schedule where communication is both criss-crossed across ranks leading to hangs due to 1. looped nature of schedules, 2. batched nature of forward + backward in 1f1b phase.

<img width="1370" alt="image" src="https://github.com/pytorch/pytorch/assets/14858254/a07c2b1d-8a99-420b-9ba3-32a0115d228b">

In the current implementation, it is difficult to fix these hangs since it requires `dist.recv` from a prior point in time, but each rank operates on its own step schedule and does not have knowledge of other ranks operations to perform the `recv` prior to their own `send`.

## New implementation

The new implementation is split into 2 parts:

1. Creating the pipeline order.

Each rank will create the timestep normalized ordering of all schedule actions across all ranks. This is created once during the initialization of the schedule class. The timestep between each rank is normalized as each rank can only have 1 computation action (forward or backward) during that timestep.

<img width="1065" alt="image" src="https://github.com/pytorch/pytorch/assets/14858254/196f2347-7ff4-49cf-903b-d8db97d1156f">

3. Executing the pipeline order.

Once the pipeline order is determined, execution is simple because as each rank will perform its send to its peer (based on whether they did forward and backward). Now that each rank has a global understanding of the schedule, they can check their previous and next neighbor ranks to see if they need to recv any activations/gradients from them. Therefore, during execution, each rank is aligned and executing the same time step.

## Benefits

- Implementation is faster since 1f1b computation can now be split up in two time steps, 1 for forward and 1 for backward.
- Debugging is easier since we can now determine which timestep each rank is hung on
- Testing is easier since we can just validate the pipeline order, without running the schedule. This allows us to test on large amount of ranks without actually needing the GPUs.




cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k 

[ghstack-poisoned]

H-Huang committed:
Update on "[pipelining] rewrite interleaved 1f1b"

## Context

Interleaved 1F1B has multiple points in the schedule where communication is both criss-crossed across ranks leading to hangs due to 1. looped nature of schedules, 2. batched nature of forward + backward in 1f1b phase.

<img width="1370" alt="image" src="https://github.com/pytorch/pytorch/assets/14858254/a07c2b1d-8a99-420b-9ba3-32a0115d228b">

In the current implementation, it is difficult to fix these hangs since it requires `dist.recv` from a prior point in time, but each rank operates on its own step schedule and does not have knowledge of other ranks operations to perform the `recv` prior to their own `send`.

## New implementation

The new implementation is split into 2 parts:

1. Creating the pipeline order.

Each rank will create the timestep normalized ordering of all schedule actions across all ranks. This is created once during the initialization of the schedule class. The timestep between each rank is normalized as each rank can only have 1 computation action (forward or backward) during that timestep.

<img width="1065" alt="image" src="https://github.com/pytorch/pytorch/assets/14858254/196f2347-7ff4-49cf-903b-d8db97d1156f">

3. Executing the pipeline order.

Once the pipeline order is determined, execution is simple because as each rank will perform its send to its peer (based on whether they did forward and backward). Now that each rank has a global understanding of the schedule, they can check their previous and next neighbor ranks to see if they need to recv any activations/gradients from them. Therefore, during execution, each rank is aligned and executing the same time step.

## Benefits

- Implementation is faster since 1f1b computation can now be split up in two time steps, 1 for forward and 1 for backward.
- Debugging is easier since we can now determine which timestep each rank is hung on
- Testing is easier since we can just validate the pipeline order, without running the schedule. This allows us to test on large amount of ranks without actually needing the GPUs.




cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k 

[ghstack-poisoned]

wconstab commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -685,6 +697,130 @@ def __init__(
+    def _calculate_single_rank_operations(self, rank) -> List[Optional[_Action]]:
+        def get_rank_warmup_ops(rank):
+            # Warms up operations for last stage
+            warmups_ops_last_stage = (self.n_local_stages - 1) * self.pp_group_size
```
if we were to use this logic for the non-looped case (n_local_stages==1) this logic would break.  Is there a more robust formula we could use?

perhaps, computing this relative to the globally last stage first, and then just offsetting it by the local stage_id?

H-Huang replied:
I think each schedule can implement its own function for the `_calculate_single_rank_operations`, in loopedBFS this entire function is different (https://github.com/pytorch/pytorch/pull/127796/files#diff-bcea404838585952f926a03a7da41b33f8bdc541a955ef25345a69ab2e2a286eR727). What could be shared is the `step_microbatches` logic.

> perhaps, computing this relative to the globally last stage first, and then just offsetting it by the local stage_id?

I could try, but i think there may be some edge case with this

wconstab replied:
ok. better to refactor later anyway, i guess.

wconstab commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -694,161 +830,86 @@ def _step_microbatches(
+
+            # do the communication
+            if ops:
+                dist.batch_isend_irecv(ops).pop().wait()
```
why not use the batch_p2p helper? it adds useful logging

H-Huang replied:
Oops, will update it

wconstab commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -694,161 +830,86 @@ def _step_microbatches(
+                if computation_type == _ComputationType.FORWARD:
+                    # Next rank doing forward has no influence for the current rank backward recv
+                    pass
+                elif computation_type == _ComputationType.BACKWARD:
```
_configure_data_parallel_mode is missing and needs to be added back

wconstab replied:
we should add a test to our schedule tests that checks if configure_data_parallel is called properly. (another PR)

H-Huang replied:
Thanks, good catch. I added it back in

H-Huang replied:
Will follow up with a test for it too

wconstab commented on torch/distributed/pipelining/PipelineSchedule.py:
```diff
@@ -694,161 +830,86 @@ def _step_microbatches(
+        stage_index_to_stage: Dict[int, _PipelineStageBase] = {
+            stage.stage_index: stage for stage in self._stages
+        }
+        prev_rank: int = (self.rank - 1) % self.pp_group_size
```
it looks like this schedule assumes there will be no skip connections.  If so, it needs to be asserted somewhere (or if it is maybe I missed it).

H-Huang replied:
Yeah the schedule assumes no skip connections. Checking / asserting for skip connections may be a bit more involved, I can do it in a follow up PR

H-Huang replied:
Fix needed in `ManualPipelineStage`

H-Huang replied:
follow up with an issue

H-Huang replied:
https://github.com/pytorch/pytorch/issues/127964

H-Huang committed:
Update on "[pipelining] rewrite interleaved 1f1b"

## Context

Interleaved 1F1B has multiple points in the schedule where communication is both criss-crossed across ranks leading to hangs due to 1. looped nature of schedules, 2. batched nature of forward + backward in 1f1b phase.

<img width="1370" alt="image" src="https://github.com/pytorch/pytorch/assets/14858254/a07c2b1d-8a99-420b-9ba3-32a0115d228b">

In the current implementation, it is difficult to fix these hangs since it requires `dist.recv` from a prior point in time, but each rank operates on its own step schedule and does not have knowledge of other ranks operations to perform the `recv` prior to their own `send`.

## New implementation

The new implementation is split into 2 parts:

1. Creating the pipeline order.

Each rank will create the timestep normalized ordering of all schedule actions across all ranks. This is created once during the initialization of the schedule class. The timestep between each rank is normalized as each rank can only have 1 computation action (forward or backward) during that timestep.

<img width="1065" alt="image" src="https://github.com/pytorch/pytorch/assets/14858254/196f2347-7ff4-49cf-903b-d8db97d1156f">

3. Executing the pipeline order.

Once the pipeline order is determined, execution is simple because as each rank will perform its send to its peer (based on whether they did forward and backward). Now that each rank has a global understanding of the schedule, they can check their previous and next neighbor ranks to see if they need to recv any activations/gradients from them. Therefore, during execution, each rank is aligned and executing the same time step.

## Benefits

- Implementation is faster since 1f1b computation can now be split up in two time steps, 1 for forward and 1 for backward.
- Debugging is easier since we can now determine which timestep each rank is hung on
- Testing is easier since we can just validate the pipeline order, without running the schedule. This allows us to test on large amount of ranks without actually needing the GPUs.




cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k 

[ghstack-poisoned]

wconstab reviewed (approved):
lgtm, please file issues for the 2 followups we identified

H-Huang commented:
@pytorchbot merge

H-Huang commented:
@pytorchbot merge -f "mac os build failure unrelated"
